lib/galaxy/managers/datasets.py

```diff
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         # ...?
         return unwrapped_val
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and context['trans'].app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            if managers.hdas.HDAManager(self.app).ok_to_edit_metadata(item.dataset_id):
```

Comment:
```suggestion
            if DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
```

Or `ok_to_edit_metadata()` should actually be a method of the `DatasetInstance` class in `lib/galaxy/model/__init__.py`, I think.